drivers/flash/flash_stm32.c

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ * Copyright (c) 2017 BayLibre, SAS.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <string.h>
+#include <flash.h>
+#include <init.h>
+#include <soc.h>
+
+#include <flash_stm32.h>
+
+#define STM32_FLASH_TIMEOUT	((u32_t) 0x000B0000)
+
+int flash_stm32_check_status(struct flash_stm32_priv *p)
+{
+#if defined(CONFIG_SOC_SERIES_STM32F4X)
+	struct stm32f4x_flash *regs = p->regs;
+#elif defined(CONFIG_SOC_SERIES_STM32L4X)
+	struct stm32l4x_flash *regs = p->regs;
+#endif
+
+	u32_t const error =
+		FLASH_FLAG_WRPERR |
+		FLASH_FLAG_PGAERR |
+#if defined(FLASH_FLAG_RDERR)
+		FLASH_FLAG_RDERR  |
+#endif
+#if defined(FLASH_FLAG_PGPERR)
+		FLASH_FLAG_PGPERR |
+#endif
+		FLASH_FLAG_PGSERR |
+		FLASH_FLAG_OPERR;
+
+	if (regs->sr & error) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int flash_stm32_wait_flash_idle(struct flash_stm32_priv *p)
+{
+#if defined(CONFIG_SOC_SERIES_STM32F4X)
+	struct stm32f4x_flash *regs = p->regs;
+#elif defined(CONFIG_SOC_SERIES_STM32L4X)
+	struct stm32l4x_flash *regs = p->regs;
+#endif
+	u32_t timeout = STM32_FLASH_TIMEOUT;
+	int rc;
+
+	rc = flash_stm32_check_status(p);
+	if (rc < 0) {
+		return -EIO;
+	}
+
+	while ((regs->sr & FLASH_SR_BSY) && timeout) {
+		timeout--;
+	}
+
+	if (!timeout) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+void flash_stm32_flush_caches(struct flash_stm32_priv *p)
+{
+#if defined(CONFIG_SOC_SERIES_STM32F4X)
+	struct stm32f4x_flash *regs = p->regs;
+#elif defined(CONFIG_SOC_SERIES_STM32L4X)
+	struct stm32l4x_flash *regs = p->regs;
+#endif
+
+	if (regs->acr.val & FLASH_ACR_ICEN) {
+		regs->acr.val &= ~FLASH_ACR_ICEN;
+		regs->acr.val |= FLASH_ACR_ICRST;
+		regs->acr.val &= ~FLASH_ACR_ICRST;
+		regs->acr.val |= FLASH_ACR_ICEN;
+	}
+
+	if (regs->acr.val & FLASH_ACR_DCEN) {
+		regs->acr.val &= ~FLASH_ACR_DCEN;
+		regs->acr.val |= FLASH_ACR_DCRST;
+		regs->acr.val &= ~FLASH_ACR_DCRST;
+		regs->acr.val |= FLASH_ACR_DCEN;
+	}
+}
+
+static int flash_stm32_read(struct device *dev, off_t offset, void *data,
+	size_t len)
+{
+	if (!flash_stm32_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	memcpy(data, (void *) CONFIG_FLASH_BASE_ADDRESS + offset, len);
+
+	return 0;
+}
+
+int flash_stm32_erase(struct device *dev, off_t offset, size_t len)
+{
+	struct flash_stm32_priv *p = dev->driver_data;
+	int i, rc = 0;
+
+	if (!flash_stm32_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	k_sem_take(&p->sem, K_FOREVER);
+
+#if defined(CONFIG_STM32_FLASH_ERASE_SECTOR)
+	i = stm32_get_sector(offset);
+	for (; i <= stm32_get_sector(offset + len - 1); i++) {
+		rc = flash_stm32_erase_sector(i, p);
+		if (rc < 0) {
+			break;
+		}
+	}
+#elif defined(CONFIG_STM32_FLASH_ERASE_PAGE)
+	i = flash_stm32_get_page(offset);
+	for (; i <= flash_stm32_get_page(offset + len - 1) ; ++i) {
+		rc = flash_stm32_erase_page(i, p);
+		if (rc < 0) {
+			break;
+		}
+	}
+#else
```

Comment:
I should admit the result is not as simple as I'd have expected, specially if we need to define additional CONFIG_ switches.
What about introducing a new generic function flash_stm32_block_erase_loop(offset, p, len)?
Then, you can have get/erase_page and get/erase_sector in respective f4 and l4 files.